Enable `@edit @time 1+1`, fix #3377

dhoegh opened the pull request:
This allows for writing ~~`methods(:@time)`, `edit(:@time)`  and~~ `@edit @time 1+1` plus the equivalent for `which` and `less`. ~~`methods` can also check for macro methods as  `methods(:@time, Tuple{Any})`~~. The `@edit`, `@less` and `@which` do also check for method specialization of the macro.

dhoegh committed:
Fix #3377. It allows for writing `@methods @time 1+1`, `@functionloc @time
1+1` plus the equivalent for `@which`, `@edit` and `@less`. The macros check for method specialization of the macro and return the dispatched macro. The `@functionloc` macro is exported and the documentation is updated.

StefanKarpinski commented:
:100:

yuyichao commented on base/interactiveutil.jl:
```diff
@@ -254,8 +254,14 @@ function gen_call_with_extracted_types(fcn, ex0)
-    if !isa(ex, Expr)
+    if fcn in [:which, :less, :edit] && ex0.head == :macrocall
+        # special case @edit @time 1+2 to edit the macro
+        is_macro = true
```
Why is the special case needed?

dhoegh replied:
The special case is needed because when the `expand` function called above expands the macro/insert its code into the expression, and hence the `edit` call will call edit on the expanded code. The special casing makes `which`, `less` and `edit` macro call the macrocall instead of the code the macro expands to. This enables the user to write: `@edit @time 1+1`, this will go to the `@time` macro's definition.

yuyichao replied:
Is there a case you don't want to use operate on the macro?

yuyichao replied:
My main concern is that this doesn't seem to be the way this should work. If it is somehow really necessary to have different behavior for different macros that use this function, it should be controlled explicitly with a optional/keyword argument to this function instead.

dhoegh replied:
I understand your concern, but from my point of view it do not make sense to ask for `@edit @random_macro max(1,1)` and not point to the macro. As a macro can generate arbitrary code for you, and hence you do not know what you call `@edit` on.

tkelman commented on base/reflection.jl:
```diff
@@ -170,6 +170,21 @@ end
+    end
+    fn
+end
+_get_macro(fn, sym::Symbol) = fn.(sym)
```
should probably use getfield for these, this syntax might be changing meaning

dhoegh commented:
I have pushed a new commit which adresses the comments. I did also find a bug, caused by wrong `esc`. I have addressed it and changed the tests to test it. Ping me when it is ready for a merge, then i will squas the commits.

dhoegh commented:
I have squased the commits.

JeffBezanson commented on base/reflection.jl:
```diff
@@ -179,6 +194,10 @@ function methods(f::ANY,t::ANY)
 function _methods(f::ANY,t::ANY,lim)
+    #special case Expr to allow methods(@time,Tuple{Any})
+    if _is_empty_macrocall(f)
+        f = _get_macro(current_module(),f.args[1])
```
This code is incorrect --- it makes it impossible for the system to query the methods of an `Expr` object. Exprs don't have any methods of course, but that just means this function must return an empty array.

dhoegh replied:
I do not quite follow you. The meaning of the function is to get function object of the macro from the `getfield` expression. Do you have an alternative way of obtaining the result?

JeffBezanson replied:
You can't do this in `_methods`. `_methods(f, ...)` must return the list of methods you could call by doing `f(...)` *with the specific object* `f`. The following (current) behavior is correct:

```
julia> e = Expr(:macrocall, symbol("@time"))
:(@time)

julia> e()
ERROR: MethodError: objects of type Expr are not callable
 in eval(::Module, ::Any) at ./boot.jl:275

julia> methods(e)
# 0 methods for generic function "(::Expr)":
```

JeffBezanson commented:
This functionality is great, but we need to figure out a different way to provide it. This implementation works via a pun, where `methods` "lies" about what the answer is for certain Exprs.

dhoegh commented:
If there was a way to obtain the function object of the macro, then it would be easier, to solve correctly. Because `Base.(Symbol("@time"))` feels awkward.

JeffBezanson commented:
I think the best approach for now is to limit this to user-interface-level features like `@edit` and `@which`. We could even add `@methods`.

dhoegh commented:
@JeffBezanson I have changed the PR to suite your comments. I have added `@methods` and `@functionloc` macro. The only hatch by adding `@methods @boundscheck` do not return all methods for `@boundscheck` but return empty as it check for a zero argument method of `@boundscheck`. Hence `@methods` and `@which` is strictly speaking  returning the same thing. The only difference is the `@methods` return a `Array{Any,1}` with one element. An alternativ approach would be to make `@methods` return all methods and if an argument is given throw an error and pointing to `@which`.
```
m = getfield(current_module(), Symbol("@boundscheck"))
methods(m) == @methods @boundscheck
@methods @boundscheck 1 # throw an error.
```

dhoegh commented:
I have added docs to `@functionloc` and edited docs for `@edit`, `@less` and `@which`. I have moved the docs inline. The commit do also removed the `@methods` macro as it provided almost the same as `@which`.

yuyichao commented on base/interactiveutil.jl:
```diff
@@ -254,8 +254,13 @@ function gen_call_with_extracted_types(fcn, ex0)
+    is_macro = false
     ex = expand(ex0)
-    if !isa(ex, Expr)
+    if fcn in [:which, :less, :edit, :functionloc] && ex0.head == :macrocall
```
I think you should check `isa(ex0, Expr)` here.

Also, I still don't understand why are the special cases necessary here. Why do you want `@code_lowered @time 1 + 2` to error instead of showing the ast?

dhoegh replied:
I see where you are going now:). I have removed the special casing of `@edit` etc. The Ast displayed by `@code_warntype` is pretty ugly, but it displays.
```
julia> @code_warntype @time 1+1
Variables:
  #self#::Base.#@time
  ex::Expr
  ####args#787#8911::TUPLE

Body:
  begin  # util.jl, line 152:
      return (top(_expr))(:block,$(Expr(:copyast, :($(QuoteNode(:( # util.jl, line 153:)))))),$(Expr(:copyast, :($(QuoteNode(:(local stats = gc_num())))))),$(Expr(:copyast, :($(QuoteNode(:( # util.jl, line 154:)))))),$(Expr(:copyast, :($(QuoteNode(:(local elapsedtime = time_ns())))))),$(Expr(:copyast, :($(QuoteNode(:( # util.jl, line 155:)))))),(top(_expr))(:local,(top(_expr))(:(=),:val,(Core._expr)(:escape,ex::Expr)::ANY)::Expr)::Expr,$(Expr(:copyast, :($(QuoteNode(:( # util.jl, line 156:)))))),$(Expr(:copyast, :($(QuoteNode(:(elapsedtime = time_ns() - elapsedtime)))))),$(Expr(:copyast, :($(QuoteNode(:( # util.jl, line 157:)))))),$(Expr(:copyast, :($(QuoteNode(:(local diff = GC_Diff(gc_num(),stats))))))),$(Expr(:copyast, :($(QuoteNode(:( # util.jl, line 158:)))))),$(Expr(:copyast, :($(QuoteNode(:(time_print(elapsedtime,diff.allocd,diff.total_time,gc_alloc_count(diff)))))))),$(Expr(:copyast, :($(QuoteNode(:( # util.jl, line 160:)))))),:val)::Expr
  end::Expr
```

yuyichao replied:
Trust me, it's not going to be "uglier" than the AST of this function (`gen_call_with_extracted_types`) itself.

tkelman commented on base/interactiveutil.jl:
```diff
@@ -295,7 +300,80 @@ for fname in [:which, :less, :edit, :code_typed, :code_warntype,
+to a variable, it returns the module in which the variable was bound. It calls out to the
+`which` function.
+"""
+:@which
```
run genstdlib (`make docs`) and commit the updates

dhoegh commented:
I will do. I will squash the commit away, when I have updated genstdlib.jl

dhoegh commented:
I have executed `genstdlib.jl` and squased the commits. This should be ready to merge if no further comment are made.

tkelman commented on base/interactiveutil.jl:
```diff
@@ -295,7 +299,80 @@ for fname in [:which, :less, :edit, :code_typed, :code_warntype,
+    @functionloc
+
+Applied to a function or macro call, it evaluates the arguments to the specified call, and
+return a tuple `(filename,line)` giving the location for the method that would be called for those arguments.
```
returns

dhoegh replied:
Thank you. I have fixed it and updated the PR.

dhoegh commented:
Bump, I have rebased the PR to resolve merge conflicts.

dhoegh commented:
Bump/any further comment's?

StefanKarpinski commented:
@JeffBezanson, you're the one who had objections before, so please take a look and merge if it's ok.

dhoegh commented:
Bump.

yuyichao commented:
LGTM. @JeffBezanson ?

yuyichao commented:
@dhoegh I think I'll merge this if you can rebase again and no one have further comment.

dhoegh commented:
@yuyichao thank you. I have rebased the PR and it should be good to merge when the CI is finished.

yuyichao commented:
.... Somehow the commit was merged but @github thinks the PR is still open and now with conflicts ..... Manually close since the merge is done....
